File: src/main/java/gruvexp/bbminigames/api/ability/AbilityTrigger.java
package gruvexp.bbminigames.api.ability;

import org.bukkit.event.entity.AreaEffectCloudApplyEvent;
import org.bukkit.event.entity.LingeringPotionSplashEvent;
import org.bukkit.event.entity.ProjectileHitEvent;

public interface AbilityTrigger {
    interface OnEntityPlace {
        void trigger(AbilityContext.EntityPlace ctx);
    }
    interface OnMelee {
        void trigger(AbilityContext.Melee ctx);
    }
    interface OnLaunch {
        void onLaunch(AbilityContext.Launch ctx);
    }
    interface OnProjectileHit {
        void onHit(ProjectileHitEvent e);
    }
    interface OnLingeringPotionUse {
        void onSplash(LingeringPotionSplashEvent e);
        void onCloudApply(AreaEffectCloudApplyEvent e);
    }
    interface OnBlockPlace {
        void onPlace(AbilityContext.BlockPlace ctx);
    }
}

File: src/main/java/gruvexp/bbminigames/api/ability/AbilityTrigger.kt
package gruvexp.bbminigames.api.ability

import gruvexp.bbminigames.api.ability.AbilityContext.*
import org.bukkit.event.entity.AreaEffectCloudApplyEvent
import org.bukkit.event.entity.LingeringPotionSplashEvent
import org.bukkit.event.entity.ProjectileHitEvent

interface AbilityTrigger {
    interface OnEntityPlace {
        fun trigger(ctx: EntityPlace)
    }

    interface OnMelee {
        fun trigger(ctx: Melee)
    }

    interface OnLaunch {
        fun onLaunch(ctx: Launch)
    }

    interface OnProjectileHit {
        fun onHit(e: ProjectileHitEvent)
    }

    interface OnLingeringPotionUse {
        fun onSplash(e: LingeringPotionSplashEvent)
        fun onCloudApply(e: AreaEffectCloudApplyEvent)
    }

    interface OnBlockPlace {
        fun onPlace(ctx: BlockPlace)
    }
}
